cluster-agent/controllers/argoproj.io/namespace_reconciler.go

```diff
@@ -0,0 +1,335 @@
+package argoprojio
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	appv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/go-logr/logr"
+	"github.com/redhat-appstudio/managed-gitops/backend-shared/apis/managed-gitops/v1alpha1"
+	"github.com/redhat-appstudio/managed-gitops/backend-shared/config/db"
+	cache "github.com/redhat-appstudio/managed-gitops/backend-shared/config/db/util"
+	"github.com/redhat-appstudio/managed-gitops/backend-shared/util"
+	sharedutil "github.com/redhat-appstudio/managed-gitops/backend-shared/util"
+	appEventLoop "github.com/redhat-appstudio/managed-gitops/backend/eventloop/application_event_loop"
+	"github.com/redhat-appstudio/managed-gitops/backend/util/fauxargocd"
+	"github.com/redhat-appstudio/managed-gitops/cluster-agent/controllers"
+	"gopkg.in/yaml.v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	appRowBatchSize             = 50               // Number of rows needs to be fetched in each batch.
+	namespaceReconcilerInterval = 30 * time.Minute // Interval in Minutes to reconcile workspace/namespace.
+	sleepIntervalsOfBatches     = 1 * time.Second  // Interval in Millisecond between each batch.
+)
+
+// This function iterates through each Workspace/Namespace present in DB and ensures that the state of resources in Cluster is in Sync with DB.
+func (r *ApplicationReconciler) StartNamespaceReconciler() {
+	r.startTimerForNextCycle()
+}
+
+func (r *ApplicationReconciler) startTimerForNextCycle() {
+	go func() {
+		// Timer to trigger Reconciler
+		timer := time.NewTimer(time.Duration(namespaceReconcilerInterval))
+		<-timer.C
+
+		ctx := context.Background()
+		log := log.FromContext(ctx).WithValues("component", "namespace-reconciler")
+
+		_, _ = util.CatchPanic(func() error {
+			runNamespaceReconcile(ctx, r.DB, r.Client, log)
+			return nil
+		})
+
+		// Kick off the timer again, once the old task runs.
+		// This ensures that at least 'namespaceReconcilerInterval' time elapses from the end of one rum to the beginning of another.
+		r.startTimerForNextCycle()
+	}()
+
+}
+
+func runNamespaceReconcile(ctx context.Context, dbQueries db.DatabaseQueries, client client.Client, log logr.Logger) {
+
+	// Fetch list of ArgoCD applications to be used later
+	processedApplicationIds := make(map[string]string)
```

Comment:
Nit: Since we don't use the value of this map anywhere, can we change the type to `bool`? 
```suggestion
	processedApplicationIds := make(map[string]bool)
```

Reply:
Yes It can be anything, values of map are is not useful in current situation, so it can be of any type.
Just for curiosity, is there any reason for choosing bool?

Reply:
I agree in this case it doesn't matter much. I prefer using bool if we are only checking if a key exists and don't care about the value. So I suggested it as a Nit. Please feel free to ignore

Reply:
FWIW I tend to do something like this: an `interface{}` for the value, then a comment explaining the use:
 ```
 // map: applications IDs seen (string) -> (map value not used)
processedApplicationIDs := make(map[string]interface{})
 ```
 But what y'all have here is fine. In Java we have `HashSet`, but no built-in equivalent in Go AFAIK :stuck_out_tongue: